draft-ietf-quic-transport.md

```diff
@@ -2063,6 +2087,20 @@ Application Error Code:
   {{app-error-codes}}).
 
 
+
+## PONG Frame {#frame-pong}
+
+The PONG frame (type=0x0d) is sent in response to a PING frame that contains
+data.  Its format is identical to the PING frame ({{frame-ping}}).
+
+An endpoint that receives an unsolicited PONG frame - that is, a PONG frame
+containing a payload that is either empty or not identical to the content of a
+PING frame that the endpoint previously sent - MUST generate a connection error
+of type UNSOLICITED_PONG.  Note that the Data field of the frame is the only
```

Comment:
You'll need to account for retransmission/duplication of PONG packets. If an endpoint transmits a PONG, has a timeout and retransmits the PONG, then the receiver of the first PONG will discard state associated with the PING, and will close the connection on receiving the second PONG.

You could ask the PING sender to remember PINGs sent over the past X RTTs, or simply discard PONGs with unknown data. I'd recommend (and implement) the latter, since its easier, has very little state, and there's no way to enforce that an endpoint terminate a connection on receiving a bad PONG.

Reply:
...or you could include a MAC in the payload and avoid all state, thereby avoiding the problem.

Reply:
I don't understand. My suggestion here was to remove the requirement that a PONG MUST match an open PING. This seems unnecessary.

Reply:
Would you be OK with MAY?  I want to ensure that bad actors can't guess PONG values without consequence.

Reply:
Bad actors can simply send random PONGs and cause the connection to be torn down, that's quite a bad consequence, isn't it? I was trying to prevent the connection from getting torn down because of legitimate operational issues, such as a duplicate PONG reaching the PING sender late. Basically, you're currently requiring that the sender of a PING hold some state without stating exactly what state, and if it sees a PONG that doesn't match the unspecified state, that it kill the connection. 

I am fine with saying MAY, but I would qualify the sentence with something like "If the receiver of a PONG can be certain that the received Data was not sent by it .... ". WDYT?

Reply:
I can work with that; it's what I was thinking.

FWIW, I didn't have a design that required state in mind.  But you are right to point out that a MUST narrows the range of reasonable designs considerably.

Not sure about the "bad actors" point you make here.  The point is to have the connection torn down if someone starts generating random PONGs.  After all, if you want to close the connection and you have the keys, you have other more direct methods available to you.

Reply:
I can totally see that you can build this without state, but we have to prepare for a naive implementer who doesn't recognize that a simple implementation that remembers exactly one PING is likely to break in a variety of situations.

About the "bad actors", agreed.